include/swift-c/DependencyScan/DependencyScan.h

```diff
@@ -497,6 +524,69 @@ SWIFTSCAN_PUBLIC swiftscan_string_ref_t swiftscan_compute_cache_key(
     swiftscan_cas_t cas, int argc, const char **argv, const char *input,
     swiftscan_output_kind_t kind, swiftscan_string_ref_t *error);
 
+/// Query the result of the compilation using the output cache key. \c globally
+/// suggests if the lookup should check remote cache if such operation exists.
+/// If remote cache is checked and result is found, the output is materialized
+/// locally so that later replay action should not require remote actions.
+/// If the return value is SWIFTSCAN_CACHE_RESULT_ERROR, the error message is
+/// returned via \c error parameter and it caller needs to free the message
+/// using `swiftscan_string_dispose`.
+SWIFTSCAN_PUBLIC swiftscan_cache_result_t
+swiftscan_cache_query(swiftscan_cas_t cas, const char *key, bool globally,
+                      swiftscan_string_ref_t *error);
+
+/// Async version of `swiftscan_cache_query` where result is returned via
+/// callback. \c ctx is an opaque value that passed to the callback and
+/// \c swiftscan_cache_cancellation_token_t will return an token that can be
+/// used to cancel the async operation. The token needs to be freed by caller
+/// using `swiftscan_cache_cancellation_token_dispose`.
+SWIFTSCAN_PUBLIC void swiftscan_cache_query_async(
+    swiftscan_cas_t cas, const char *key, bool globally, void *ctx,
+    void (*callback)(void *ctx, swiftscan_cache_result_t,
+                     swiftscan_string_ref_t error),
+    swiftscan_cache_cancellation_token_t *);
+
+/// Make the cache compilation available globally. \c callback will be called
+/// on completion.
+/// \c swiftscan_cache_cancellation_token_t will return an token that can be
+/// used to cancel the async operation. The token needs to be freed by caller
+/// using `swiftscan_cache_cancellation_token_dispose`.
+SWIFTSCAN_PUBLIC void swiftscan_cache_make_global_async(
+    swiftscan_cas_t cas, const char *key, void *ctx,
+    void (*callback)(void *ctx, swiftscan_string_ref_t error),
+    swiftscan_cache_cancellation_token_t *);
+
+/// Cancel the async cache action that is associated with token.
+SWIFTSCAN_PUBLIC void
+    swiftscan_cache_action_cancel(swiftscan_cache_cancellation_token_t);
+
+/// Dispose the cancellation token.
+SWIFTSCAN_PUBLIC void swiftscan_cache_cancellation_token_dispose(
+    swiftscan_cache_cancellation_token_t);
+
+/// Replay the cached compilation using command-line.
+/// If the return value is SWIFTSCAN_CACHE_RESULT_ERROR, the error message is
+/// returned via \c error parameter and it caller needs to free the message
+/// using `swiftscan_string_dispose`.
+SWIFTSCAN_PUBLIC swiftscan_cache_result_t swiftscan_cache_replay_compilation(
```

Comment:
I was expecting this to have a "key" parameter and use that.  Instead it looks like it is triggering a full compilation if the result is not cached.

Reply:
swift compilation has more than one `key`. I have a patch for swift driver to compute all the keys already and driver/build system can query before calling replay. I actually would like this API to not compile at all, only replay. So preferably, only cache key goes into this API, but because all the limitation mentioned above, it is using command-line to reconstruct the keys and all the options for replay.